src/main/java/org/openstreetmap/atlas/utilities/collections/OptionalIterable.java

```diff
@@ -31,17 +32,24 @@ public OptionalIterable(final Iterable<Optional<T>> iterable)
             @Override
             public boolean hasNext()
             {
-                while (this.iterator.hasNext())
+                if (previousElement.isPresent())
                 {
-                    final Optional<T> current = this.iterator.next();
-                    if (current.isPresent())
+                    return true;
+                }
+                else
+                {
+                    while (this.iterator.hasNext())
                     {
-                        this.previousElement = current;
-                        return true;
+                        final Optional<T> current = this.iterator.next();
+                        if (current.isPresent())
+                        {
+                            this.previousElement = current;
+                            return true;
+                        }
+                        this.previousElement = Optional.empty();
```

Comment:
Nit: Do we need to set `previousElement` to empty here? It's already initialized to empty and if it's not present, we'd just be setting it again.

Reply:
logically this assignment is pointless so I'll remove it.